### ISSUE TITLE
gradle fails immediately with JDK10.

### ISSUE THREAD

**Issue by prrace** [gradle/gradle#2454]
gradle does not run at all with JDK 10 since it tries, and fails to recognise the java.version property value.

### Expected Behavior
As much as possible gradle should not care or fail if it does not recognise the version
Where that does not work, gradle should accept unrecognised versions as being
"newer" than the latest it knows and supporting all features.
Failing that gradle should provide a flag, or system property, to allow a user to tell
it to accept it as being some version it will accept.

### Current Behavior
$ java -version
java version "10-internal"
Java(TM) SE Runtime Environment (build 10-internal+14-jdk-1014)
Java HotSpot(TM) 64-Bit Server VM (build 10-internal+14-jdk-1014, mixed mode)

$ gradle-4.0.1/bin/gradle -v

FAILURE: Build failed with an exception.

* What went wrong:
Could not determine java version from '10-internal'.

Looking at the implementation in src/base-services/org/gradle/api/JavaVersion.java
which is an enum class, it probably successfully parses the version as "10" but this is
then used to index in to the array of enum values .. which has only 9 values defined
from VERSION_1_1(false) ... VERSION_1_9(true);
Thus it rejects the version and throws an exception :
        throw new IllegalArgumentException(String.format("Could not determine java version from '%s'.", name));

I think this approach /design needs to change as it is one of perpetual catch-up.

### Context
It is a major problem since we use gradle in building parts of JDK 10 itself ! ie the JavaFX APIs
It is causing our build+test for JavaFX 10 to fail.

### Steps to Reproduce
Download openjdk 10 sources + build + use to run gradle.

**Comment by AlanBateman** [comment]
See also JEP 223 (http://openjdk.java.net/jeps/223) where the version scheme is defined.

**Comment by melix** [comment]
Thanks @prrace . I agree that Gradle should do best effort not to fail when it doesn't recognize a version. There are however lots of parts of the code that _effectively_ depend on the runtime version (and don't get me started with jigsaw ;)).

One thing we can do is to add a dummy "Unknown" version in the enum for such cases, and we would assume that the version is the latest known (so here, it would assume behavior of Java 9).

Or, we can just add 10 as a target (and yes, the enum value would be the awkward 1_10, but well...). I'll just submit a PR to add support for 10, then we can unblock you.

BTW, glad to know Gradle is used to build parts of the JDK, now you only have to expand it to the native parts ;)

**Comment by AlanBateman** [comment]
@melix - the Runtime.version() API is useful to know about as it avoids the need to parse `java.version`. It's a new API in Java SE 9 so means using reflection or a MR JAR if you are building on older versions but important to know that this is the API to use going forward. It's probably better to treat a the major version > 9 as "newer than 9" rather than committing to a still-early-in-development-JDK-10.

PS: Gradle is to use to build JavaFX.

**Comment by melix** [comment]
Thanks for letting us know about `Runtime.version()`. The problem is always that we cannot statically link to this API, so we need to resort on reflection or method handles to do this.

**Comment by prrace** [comment]
I think that essentially "unknown" should be equated to "latest and greatest" and it is better to
do that than just add 1_10 to the enum.
Until you know about a new version and some particular problem with it, it is better to assume
everything that you relied on and expected in version N, will be fine in N+M .. where M > 0.

**Comment by prrace** [comment]
FWIW I tried gradle 4.0.1 with JDK 9 to see if there are any other blockers that would make
this fix moot. Sadly there is !
We already noticed it with gradle 3.4 and are blocked from even moving to that.
See https://bugs.openjdk.java.net/browse/JDK-8177778
I don't know if a gradle bug was filed on that but I'll try to find out.

**Comment by melix** [comment]
Gradle 4.0.1 is not compatible with JDK 9. Gradle 4.1 (for which an RC is due soon) will. In the meantime, you can try with a [snapshot of 4.1](https://services.gradle.org/distributions-snapshots/gradle-4.1-20170718061434+0000-bin.zip).

**Comment by AlanBateman** [comment]
Will 4.1 parse java.version so that it can work on JDK 10 too?

**Comment by melix** [comment]
Yes, look at the linked pull request.

**Comment by prrace** [comment]
> you can try with a snapshot of 4.1.
I will do that but I filed  https://github.com/gradle/gradle/issues/2537 since
I don't see anything that indicates this problem is known .. and 4.0.1 isn't
exactly old - released just 12 days ago.

**Comment by prrace** [comment]
.. and indeed it fails the same with the 4.1 snapshot

**Comment by melix** [comment]
> I don't see anything that indicates this problem is known .. and 4.0.1 isn't exactly old - released just 12 days ago.

Age doesn't say much, absolutely no work has been done on the 4.0 branch to support JDK 9, but there has been quite a few commits on 4.1 about it (just like we started work for 4.2 weeks ago).

> .. and indeed it fails the same with the 4.1 snapshot

There's quite a bit of work to do, so I'm glad you bring #2537 to our knowledge. Please be aware, however, that we will support running on JDK 9, but fully fledge support for modules is far from complete. In the meantime, we wrote up a [guide](https://guides.gradle.org/building-java-9-modules/) that should explain how to deal with some issues.

**Comment by prrace** [comment]
I am hoping that since 3.4 broke this it is not a huge new thing to deal with .. more a case of
backing out what broke it.

**Comment by prrace** [comment]
I wonder if a 3.2.2 release .. which does work for 9 .. fixing just the JDK 10 parsing ... 
would be the more sure way to get us past the immediate problem and we can fight the other battles at a more leisurely pace.

**Comment by eriwen** [comment]
@melix Your change is merged for v4.1 so we can close this as delivered, right?

**Comment by melix** [comment]
@eriwen that's correct
@prrace I fail to see how 3.x can behave correctly on JDK 9. Unless you added a bunch of CLI arguments and disabled the daemon maybe?

**Comment by prrace** [comment]
Per our build wiki https://wiki.openjdk.java.net/display/OpenJFX/Building+OpenJFX#BuildingOpenJFX-Gradle

we needed to set 
export _JAVA_OPTIONS="-Dsun.reflect.debugModuleAccessChecks=true --add-opens=java.base/java.lang=ALL-UNNAMED --add-opens=java.base/java.util=ALL-UNNAMED --add-opens=java.base/java.lang.invoke=ALL-UNNAMED --add-opens=java.base/java.io=ALL-UNNAMED --add-opens=java.base/java.util.concurrent=ALL-UNNAMED --add-opens=java.base/java.text=ALL-UNNAMED"

**Comment by prrace** [comment]
I've looked at the code for how this was fixed :
https://github.com/gradle/gradle/commit/500485df3a1898d4637bfa42b73e356941b4af91#diff-b1bd5823f786153a225f410abe7798b4
and it appears that new developments mean it will need to be updated.
I can file a new bug but I hope the author of this fix will see this as well.

There is a proposal in JDK-land to
(1) Have "Major" releases every 6 months. 
See http://mail.openjdk.java.net/pipermail/discuss/2017-September/004281.html
This seems very likely to be accepted.
So you may expect JDK 10 in March 2018, JDK 11 in September 2018 and so forth.
The current version numbering handling in Gradle is going to be out-dated and unable
to run on any JDK more than 6 months old .. and support for old versions will be dropped
when the new one comes out

(2) The version numbering (probably) WON'T be 10, 11, etc.
It will be 18.3 (March 2018. ie MM/YY==03/18), 18.9 (September 2018, ie MM/YY==09/18) etc ..
This is also under discussion and was proposed at this
blog https://mreinhold.org/blog/forward-faster which is linked from the email above

So I think my proposal above in this thread needs to be re-visited 
> Where that does not work, gradle should accept unrecognised versions as being
> "newer" than the latest it knows and supporting all features.

NB. I estimate that if the proposals are quickly adopted, JDK 10 will morph into JDK 18.3
sometime in early November this year  and we'll be broken again !

**Comment by melix** [comment]
Thanks @prrace . I'm well aware of this proposal (and hope that the yy.x pattern will not be accepted as it will remove all visibility of major "platform changes", but it's another story). I also reckon we need to change parsing (as I discussed, coincidentally, this morning with @pioterj ). However as I understand the plan is also to upgrade the class format with each release, which also means that we would have no choice but releasing a new Gradle version for each release too (probably with a new ASM version bundled). So technically speaking, we could add versions as they are released... I don't find this very practical TBH, I would very much prefer semantic versioning.

**Comment by prrace** [comment]
Hmm .. updating the class file version is not something I'd realised was intended for every one
of these releases. It may be inevitable depending on what was added, but I don't know whether
it will be required regardless .. however even if it happens only 50% of the time it is an issue.
And if you are manipulating, even just parsing, the class files I suppose that may be an issue.

And there's a practical problem here for anyone who needs the gradle version
updated *before* the new JDK version is released. Which is our case.
But I don't remember the class file version being bumped during 9 causing us any problems.

**Comment by melix** [comment]
> Hmm .. updating the class file version is not something I'd realised was intended for every one
of these releases. It may be inevitable depending on what was added, but I don't know whether
it will be required regardless .. however even if it happens only 50% of the time it is an issue.
And if you are manipulating, even just parsing, the class files I suppose that may be an issue.

Yes, see https://twitter.com/JoroRoss/status/905507900537401344

> And there's a practical problem here for anyone who needs the gradle version
updated before the new JDK version is released. Which is our case.
But I don't remember the class file version being bumped during 9 causing us any problems.

Well it has been for us. For a long time there wasn't even an alpha version of ASM that recognized the Java 9 class format so we had to patch our class readers to workaround...

**Comment by automenta** [comment]
```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':util:compileJava'.
> java.lang.IllegalArgumentException (no error message)

Caused by: java.lang.IllegalArgumentException
        at org.objectweb.asm.ClassReader.<init>(ClassReader.java:169)
        at org.objectweb.asm.ClassReader.<init>(ClassReader.java:152)
        at org.gradle.api.internal.tasks.compile.incremental.analyzer.DefaultClassDependenciesAnalyzer.getClassAnalysis(DefaultClassDependenciesAnalyzer.java:32)
```

im overriding org.objectweb.asm.ClassReader class. it involves commenting two lines that have no consequence

```
public ClassReader(final byte[] b, final int off, final int len) {
        this.b = b;


        // checks the class version
//        if (readShort(off + 6) > Opcodes.V1_9) {
//            throw new IllegalArgumentException();
//        }
```

**Comment by melix** [comment]
We shouldn't use this issue to report _all_ possible issues on JDK 10 or it's going to be a nightmare to handle. I created https://github.com/gradle/gradle/issues/3770 to track this.
